lib/dynamoid/adapter_plugin/aws_sdk_v2.rb

```diff
@@ -84,36 +84,55 @@ def client
         @client
       end
 
-      # Puts or deletes multiple items in one or more tables
+      # Puts multiple items in one table
+      #
+      # If optional block is passed it will be called for each written batch of items, meaning once per batch.
+      # Block receives boolean flag which is true if there are some unprocessed items, otherwise false.
+      #
+      # @example Saves several items to the table testtable
+      #   Dynamoid::AdapterPlugin::AwsSdkV2.batch_write_item('table1', [{ id: '1', name: 'a' }, { id: '2', name: 'b'}])
+      #
+      # @example Pass block
+      #   Dynamoid::AdapterPlugin::AwsSdkV2.batch_write_item('table1', items) do |bool|
+      #     if bool
+      #       puts 'there are unprocessed items'
+      #     end
+      #   end
       #
       # @param [String] table_name the name of the table
       # @param [Array]  items to be processed
       # @param [Hash]   additional options
+      # @param [Proc]   optional block
       #
       # See:
       # * http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html
       # * http://docs.aws.amazon.com/sdkforruby/api/Aws/DynamoDB/Client.html#batch_write_item-instance_method
-      #
-      # TODO handle rejections because of exceeding limit for the whole request - 16 MB,
-      # item size limit - 400 KB or because provisioned throughput is exceeded
       def batch_write_item table_name, objects, options = {}
-        requests = []
-
-        objects.each_slice(BATCH_WRITE_ITEM_REQUESTS_LIMIT) do |os|
-          requests << os.map { |o| { put_request: { item: sanitize_item(o) } } }
-        end
+        items = objects.map { |o| sanitize_item(o) }
 
         begin
-          requests.each do |request_items|
-            client.batch_write_item(
+          while items.present? do
+            batch = items.take(BATCH_WRITE_ITEM_REQUESTS_LIMIT)
+            items.shift(BATCH_WRITE_ITEM_REQUESTS_LIMIT)
```

Comment:
`shift` should return the batch as well so could just combine the two lines
`batch = items.shift(BATCH_WRITE_ITEM_REQUESTS_LIMIT)`